crates/miden-lib/asm/kernels/transaction/lib/account.masm

```diff
@@ -1263,6 +1260,159 @@ export.save_account_procedure_data
+        # => [is_map_slot_type, slot_ptr, slot_idx, storage_slots_ptr]
+
+        if.true
+            # load the current map root
+            padw dup.4 mem_loadw
+            # => [MAP_ROOT, slot_ptr, slot_idx, storage_slots_ptr]
+
+            # overwrite the map root with the root of an empty SMT, so we can insert the entries of
+            # the map into an empty map and then check whether the resulting root matches MAP_ROOT.
+            exec.constants::get_empty_smt_root
+            # => [EMPTY_SMT_ROOT, MAP_ROOT, slot_ptr, slot_idx, storage_slots_ptr]
+
+            dup.8 mem_storew dropw
+            # => [MAP_ROOT, slot_ptr, slot_idx, storage_slots_ptr]
+
+            dup.5
+            # => [slot_idx, MAP_ROOT, slot_ptr, slot_idx, storage_slots_ptr]
+
+            dup.4 dup.4 dup.4 dup.4
+            # => [MAP_ROOT, slot_idx, MAP_ROOT, slot_ptr, slot_idx, storage_slots_ptr]
+
+            # add map entries to account storage, which does two important things:
+            # - insert all entries into an empty storage map, so we can validate that exactly those
+            #   entries are present that are contained in the map (no more, no less)
+            # - insert all entries into the in-kernel account storage delta
+            exec.insert_new_storage_map_entries
+            # => [MAP_ROOT, slot_ptr, slot_idx, storage_slots_ptr]
+
+            # load the root after all entries have been inserted
+            padw movup.8 mem_loadw
+            # => [CURRENT_ROOT, MAP_ROOT, slot_idx, storage_slots_ptr]
+
+            # the storage map root after inserting all entries must match the map root that was
+            # committed to as part of account creation
+            assert_eqw.err=ERR_ACCOUNT_STORAGE_MAP_ENTRIES_DO_NOT_MATCH_MAP_ROOT
+            # => [slot_idx, storage_slots_ptr]
+        else
+            drop
+            # => [slot_idx, storage_slots_ptr]
+        end
+        # => [slot_idx, storage_slots_ptr]
+
+        dup neq.0
+        # => [should_continue, slot_idx, storage_slots_ptr]
+    end
+    # => [slot_idx, storage_slots_ptr]
+
+    drop drop
+    # => []
+end
+
+#! Inserts the entries of the provided storage map root into the account.
+#!
+#! These entries must be present in the advice provider with the map root as the key. Each entry is
+#! is inserted using set_map_item on an initially empty SMT root.
+#! This allows checking whether the root of the SMT with all entries inserted matches the root of
+#! the map the account commits to.
+#! It also inserts all entries into the in-kernel delta, where the initial value of each entry will
+#! be set to the empty word so the delta for this map is computed as if the map had initially been
+#! empty.
+#!
+#! Inputs:
+#!   Operand stack: [MAP_ROOT, slot_idx]
+#!   Advice map: { MAP_ROOT: [MAP_ENTRIES] }
+#! Outputs: []
+proc.insert_new_storage_map_entries
+    adv.push_mapvaln
+    # OS => [MAP_ROOT, slot_idx]
+    # AS => [num_entries, [MAP_ENTRIES]]
```

Comment:
I wonder if the comment on line 1363 is correct: we are not getting the number of entries here, right? Rather, we are getting the number of elements in the entries, and then the number of entries can be computed as `num_elements / 8`.